### PR TITLE
Check if in a LXC container and skip Chrony

### PR DESCRIPTION
**Pull request by jaminmc**
This checks if the script is running in a LXC container like Proxmox, or LXD, and does not install Chrony, as that package fails in a container, and the host itself has correct time.

fixes Issue #119

### CODE REVIEW

File: sng_freepbx_debian_install.sh [hunk 1]
```diff
@@ -944,11 +958,18 @@ sed -i -e "s|^TFTP_DIRECTORY=\"/srv\/tftp\"$|TFTP_DIRECTORY=\"/tftpboot\"|" /etc
...
 if [ ! -f /proc/net/if_inet6 ]; then
 	sed -i -e "s|^TFTP_OPTIONS=\"--secure\"$|TFTP_OPTIONS=\"--secure --ipv4\"|" /etc/default/tftpd-hpa
-	sed -i -e "s|^DAEMON_OPTS=\"-F 1\"$|DAEMON_OPTS=\"-F 1 -4\"|" /etc/default/chrony
+    if [ $CONTAINER != true ]; then
+        sed -i -e "s|^DAEMON_OPTS=\"-F 1\"$|DAEMON_OPTS=\"-F 1 -4\"|" /etc/default/chrony
+    fi
 fi
 # Start the tftp & chrony daemons
-systemctl unmask tftpd-hpa.service chrony.service
-systemctl start tftpd-hpa.service chrony.service
```
Comment (JoseGoncalves) [inline sng_freepbx_debian_install.sh:951]: In a similar way as above, this seems better to me:
```bash
systemctl unmask tftpd-hpa.service
systemctl start tftpd-hpa.service
if [ "$CONTAINER" != true ]; then
    systemctl unmask chrony.service
    systemctl start chrony.service
fi
```

File: sng_freepbx_debian_install.sh [hunk 2]
```diff
@@ -22,14 +22,22 @@
...
-SCRIPTVER="1.12"
+SCRIPTVER="1.13"
 ASTVERSION=21
 PHPVERSION="8.2"
 LOG_FOLDER="/var/log/pbx"
 LOG_FILE="${LOG_FOLDER}/freepbx17-install-$(date '+%Y.%m.%d-%H.%M.%S').log"
 log=$LOG_FILE
 SANE_PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
 
+# Check if running in a Container
```
[the remark this answers is not in the document]
Comment (JoseGoncalves) [reply sng_freepbx_debian_install.sh:33]: Nevertheless, this check should always be done after we are sure that the script is running as root and also after the version checking, i.e., around [line 652](https://github.com/FreePBX/sng_freepbx_debian_install/blob/master/sng_freepbx_debian_install.sh#L652).

**Review by JoseGoncalves — CHANGES_REQUESTED** [review]
My improvement suggestions...

**Comment by jaminmc** [comment]
I implemented the reccomendations that @JoseGoncalves made.

**Comment by JoseGoncalves** [comment]
> I implemented the reccomendations that @JoseGoncalves made.

Hi @jaminmc. Thanks for pushing this.

Nevertheless, there are some things missing:

1. You should enclose all the references to the `$CONTAINER` variable in double quotes,  i.e., you shoud write `"$CONTAINER"`, to prevent errors parsing the script when that variable is not defined (when not running in a container).
2. You should migrate the initial container test code to after the script's version check test.

**Comment by jaminmc** [comment]
> 1. You should enclose all the references to the `$CONTAINER` variable in double quotes,  i.e., you shoud write `"$CONTAINER"`, to prevent errors parsing the script when that variable is not defined (when not running in a container).
I decided to to change the Variable to `INSTALL_CHRONY`, and also add the option `--disable-chrony` to the script.  So not the option req in issue #119 is now included. This way if someone wants to omit it in a VM, they can.  I declared it as true before the options, and that and the Container detection set it to false to disable installing Chrony.
> 2. You should migrate the initial `running in a container` test code to after the script's version check test.
Done :)

**Comment by kguptasangoma** [comment]
hey @JoseGoncalves are you good with this PR ? thanks

**Comment by JoseGoncalves** [comment]
Hi @kguptasangoma

Functionality wise, this seems fine to me and will address issues found by users that are trying to install FreePBX 17 inside a container (e.g. #119 & #123). A time sync daemon like `chrony` is not necessary inside a container, and will crash the FreePBX installation with the default container setup, due to the lack of required privileges to access system time functions.

In my view this PR has some (minor) style issues;
1. To be more in sync with the remaining options available in the script, I would replace `--disable-chrony` with `--nochrony` and use a variable named `$nochrony` instead of `$INSTALL_CHRONY` (reversing the true/false logic).
2. It uses indentation with spaces in places were tabs are used. Nevertheless, this is a generalized issue with the script (due to the contribution of several users), were in some places tabs are used and in others spaces are used. Maybe after you merge this PR you could use a tool like [shfmt](https://github.com/mvdan/sh) to clean up the script?

**Comment by jaminmc** [comment]
> My suggested styling improvements:

Fixed it. Sorry for the delay, I have no power still because of the hurricane. But I was able to fix it with my laptop.

**Review by wenzhuoz — APPROVED** [review]
It works well.

The host is an Intel NUC5, Ubuntu 24.04 (noble).

```
wenzhuo@nuc5:~$ lxc profile set freepbx limits.cpu=2 limits.memory=1GB
wenzhuo@nuc5:~$ lxc launch images:debian/12/cloud freepbx -p bridged-lan -p freepbx
Creating freepbx
Starting freepbx
wenzhuo@nuc5:~$ lxc profile show freepbx
config:
  limits.cpu: "2"
  limits.memory: 1GB
  user.network-config: |
    network:
      version: 2
      ethernets:
        eth0:
          dhcp4: false
          dhcp6: false
          addresses:
            - 172.16.10.62/24
          routes:
            - to: 0.0.0.0/0
              via: 172.16.10.1
          nameservers:
            search: [lan]
            addresses: [172.16.9.5, 172.16.9.7]
  user.user-data: |
    #cloud-config
    timezone: Asia/Shanghai
    users:
      - name: wenzhuo
        ssh_authorized_keys:
          - ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIPacv8mSeFoRftfgzFWYbl8NtEpLfQZoP+utFv+BWf9o wenzhuo@insp
        sudo: ALL=(ALL) NOPASSWD:ALL
        groups: adm, asterisk, users, www-data
        shell: /bin/bash
    apt:
      preserve_sources_list: false
      primary:
        - arches: [default]
          uri: http://mirrors.ustc.edu.cn/debian
      security:
        - arches: [default]
          uri: http://mirrors.ustc.edu.cn/debian-security
      sources_list: |
        deb $MIRROR $RELEASE main
        deb $MIRROR $RELEASE-updates main
        deb $SECURITY $RELEASE-security main
    package_upgrade: true
    packages:
      - less
      - curl
      - wget
      - openssh-server
    runcmd:
      - |
        cd /tmp
        wget http://cloud.lan/freepbx/sng_freepbx_debian_install/sng_freepbx_debian_install.sh
        bash sng_freepbx_debian_install.sh --skipversion \
          --debianmirror http://mirrors.ustc.edu.cn/debian \
          --npmmirror https://registry.npmmirror.com
description: freepbx LXD profile
devices: {}
name: freepbx
used_by:
- /1.0/instances/freepbx
```


```
wenzhuo@nuc5:~$ lxc exec freepbx bash
root@freepbx:~# grep ^Cloud-init /var/log/cloud-init-output.log
Cloud-init v. 22.4.2 running 'init-local' at Sun, 13 Oct 2024 01:41:53 +0000. Up 2.95 seconds.
Cloud-init v. 22.4.2 running 'init' at Sun, 13 Oct 2024 01:41:55 +0000. Up 4.50 seconds.
Cloud-init v. 22.4.2 running 'modules:config' at Sun, 13 Oct 2024 01:42:00 +0000. Up 9.46 seconds.
Cloud-init v. 22.4.2 running 'modules:final' at Sun, 13 Oct 2024 01:42:01 +0000. Up 11.28 seconds.
Cloud-init v. 22.4.2 finished at Sun, 13 Oct 2024 02:27:31 +0000. Datasource DataSourceNoCloud [seed=/var/lib/cloud/seed/nocloud-net
][dsmode=net].  Up 2740.85 seconds
root@freepbx:~# df -h
Filesystem                     Size  Used Avail Use% Mounted on
mypool/lxd/containers/freepbx  540G  4.9G  535G   1% /
none                           492K  4.0K  488K   1% /dev
udev                           3.9G     0  3.9G   0% /dev/zfs
efivarfs                       128K  127K     0 100% /sys/firmware/efi/efivars
tmpfs                          100K     0  100K   0% /dev/lxd
tmpfs                          100K     0  100K   0% /dev/.lxd-mounts
tmpfs                          3.9G  192K  3.9G   1% /dev/shm
tmpfs                          1.6G  208K  1.6G   1% /run
tmpfs                          5.0M     0  5.0M   0% /run/lock
root@freepbx:~# free -h
               total        used        free      shared  buff/cache   available
Mem:           953Mi       629Mi        49Mi        25Mi       274Mi       324Mi
Swap:             0B          0B          0B
```

**Comment by kguptasangoma** [comment]
Hi @jaminmc thanks a lot for your contribution, however looks like this got the conflict so could you please fix the conflict. thanks

**Comment by jaminmc** [comment]
> Hi @jaminmc thanks a lot for your contribution, however looks like this got the conflict so could you please fix the conflict. thanks

I fixed the Conflict :)

**Review by kguptasangoma — APPROVED** [review]

**Review by wenzhuoz — COMMENTED** [review]
I suggest a code alignment fix:

```diff
diff --git a/sng_freepbx_debian_install.sh b/sng_freepbx_debian_install.sh
index f6fc96f..9f10d4e 100755
--- a/sng_freepbx_debian_install.sh
+++ b/sng_freepbx_debian_install.sh
@@ -87,10 +87,10 @@ while [[ $# -gt 0 ]]; do
                        DEBIAN_MIRROR=$2
                        shift; shift # past argument
                        ;;
-    --npmmirror)
-      NPM_MIRROR=$2
-      shift; shift # past argument
-      ;;
+               --npmmirror)
+                       NPM_MIRROR=$2
+                       shift; shift # past argument
+                       ;;
                -*)
                        echo "Unknown option $1"
                        exit 1
```

**Pull request merged**